File: sections/69_master_structures_trees_mixed/problems/a_snapshot_rank/solution.cpp
#include <bits/stdc++.h>
using namespace std;
struct Node {
  int left = 0, right = 0, count = 0;
};
vector<Node> tree(1);
int add(int old, int lo, int hi, int pos) {
  int node = tree.size();
  tree.push_back(tree[old]);
  tree[node].count++;
  if (lo != hi) {
    int mid = (lo + hi) / 2;
    if (pos <= mid)
      tree[node].left = add(tree[old].left, lo, mid, pos);
    else
      tree[node].right = add(tree[old].right, mid + 1, hi, pos);
  }
  return node;
}
int kth(int before, int after, int lo, int hi, int k) {
  if (lo == hi) return lo;
  int amount = tree[tree[after].left].count - tree[tree[before].left].count,
      mid = (lo + hi) / 2;
  if (k <= amount) return kth(tree[before].left, tree[after].left, lo, mid, k);
  return kth(tree[before].right, tree[after].right, mid + 1, hi, k - amount);
}
int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);
  int maximum, q;
  if (!(cin >> maximum >> q)) return 0;
  int log = 20;
  while ((1 << log) <= q) log++;
  vector<vector<int>> up(log, vector<int>(q + 1));
  vector<int> root(q + 1), depth(q + 1);
  int versions = 0;
  auto ancestor = [&](int v, int steps) {
    for (int j = 0; j < log; j++)
      if (steps >> j & 1) v = up[j][v];
    return v;
  };
  while (q--) {
    char type;
    cin >> type;
    if (type == 'A') {
      int v, x;
      cin >> v >> x;
      ++versions;
      depth[versions] = depth[v] + 1;
      up[0][versions] = v;
      for (int j = 1; j < log; j++) up[j][versions] = up[j - 1][up[j - 1][versions]];
      root[versions] = add(root[v], 1, maximum, x);
    } else {
      int v, l, r, k;
      cin >> v >> l >> r >> k;
      int right_version = ancestor(v, depth[v] - r),
          left_version = ancestor(v, depth[v] - (l - 1));
      cout << kth(root[left_version], root[right_version], 1, maximum, k) << '\n';
    }
  }
}

File: sections/69_master_structures_trees_mixed/problems/a_snapshot_rank/solution.py
import sys


def main():
    tokens = sys.stdin.buffer.read().split()
    if not tokens:
        return
    idx = 0
    maximum = int(tokens[idx])
    q = int(tokens[idx + 1])
    idx += 2
    log = (q + 1).bit_length()
    up = [[0] * (q + 1) for _ in range(log)]
    roots = [0] * (q + 1)
    depth = [0] * (q + 1)
    left = [0]
    right = [0]
    count = [0]
    versions = 0

    def add(old, lo, hi, pos):
        node = len(count)
        left.append(left[old])
        right.append(right[old])
        count.append(count[old] + 1)
        if lo != hi:
            mid = (lo + hi) // 2
            if pos <= mid:
                left[node] = add(left[old], lo, mid, pos)
            else:
                right[node] = add(right[old], mid + 1, hi, pos)
        return node

    def ancestor(v, steps):
        bit = 0
        while steps:
            if steps & 1:
                v = up[bit][v]
            steps //= 2
            bit += 1
        return v

    def kth(before, after, k):
        lo, hi = 1, maximum
        while lo != hi:
            amount = count[left[after]] - count[left[before]]
            mid = (lo + hi) // 2
            if k <= amount:
                before = left[before]
                after = left[after]
                hi = mid
            else:
                k -= amount
                before = right[before]
                after = right[after]
                lo = mid + 1
        return lo

    out = []
    for _ in range(q):
        kind = tokens[idx].decode()
        idx += 1
        if kind == "A":
            v = int(tokens[idx])
            x = int(tokens[idx + 1])
            idx += 2
            versions += 1
            depth[versions] = depth[v] + 1
            up[0][versions] = v
            for j in range(1, log):
                up[j][versions] = up[j - 1][up[j - 1][versions]]
            roots[versions] = add(roots[v], 1, maximum, x)
        else:
            v, l, r, k = map(int, tokens[idx : idx + 4])
            idx += 4
            rv = ancestor(v, depth[v] - r)
            lv = ancestor(v, depth[v] - l + 1)
            out.append(str(kth(roots[lv], roots[rv], k)))
    print("\n".join(out))


if __name__ == "__main__":
    main()
